govuk_chat_evaluation/rag_answers/custom_deepeval/metrics/context_relevancy/context_relevancy.py

```diff
@@ -0,0 +1,202 @@
+from .schema import (
+    TruthCollection,
+    InformationNeedsCollection,
+    VerdictCollection,
+    ScoreReason,
+)
+from deepeval.errors import MissingTestCaseParamsError
+
+SchemaType = TypeVar("SchemaType", bound=BaseModel)
+
+
+class ContextRelevancyMetric(BaseMetric):
+    _required_params: List[LLMTestCaseParams] = [
+        LLMTestCaseParams.INPUT,
+        LLMTestCaseParams.ACTUAL_OUTPUT,
+    ]
+
+    def __init__(
+        self,
+        threshold: float = 0.8,
+        model: str | DeepEvalBaseLLM | None = None,
+        include_reason: bool = True,
+        strict_mode: bool = False,
+        verbose_mode: bool = False,
+        evaluation_template: Type[ContextRelevancyTemplate] = ContextRelevancyTemplate,
+    ):
+        self.threshold = 1 if strict_mode else threshold
+        self.model, self.using_native_model = initialize_model(model)
+        self.evaluation_model = self.model.get_model_name()
+        self.include_reason = include_reason
+        self.strict_mode = strict_mode
+        self.verbose_mode = verbose_mode
+        self.evaluation_template = evaluation_template
+
+    async def a_measure(
+        self,
+        test_case: LLMTestCase,
+        _show_indicator: bool = True,
+        _in_component: bool = False,
+    ) -> float:
+        check_llm_test_case_params(test_case, self._required_params, self)
+        if (
+            test_case.additional_metadata is None
+            or test_case.additional_metadata.get("structured_context") is None
+        ):
+            raise MissingTestCaseParamsError(
+                "additional_metadata['structured_context']"
+                " cannot be None for ContextRelevancyMetric."
+            )
+
+        with metric_progress_indicator(
+            self,
+            async_mode=True,
+            _show_indicator=_show_indicator,
+            _in_component=_in_component,
+        ):
+            if isinstance(test_case.additional_metadata, dict):
+                structured_context = test_case.additional_metadata.get(
+                    "structured_context", []
+                )
+                retrieval_context = [
+                    ctx.to_flattened_context_content() for ctx in structured_context
+                ]
+            else:
+                retrieval_context = []
+
+            self.evaluation_cost = 0.0
+            truth_collection: TruthCollection = await self._generate_truths(
+                retrieval_context
+            )
+            information_needs_collection: InformationNeedsCollection = (
+                await self._generate_information_needs(test_case.input)
+            )
+            verdict_collection: VerdictCollection = await self._generate_verdicts(
+                information_needs_collection, truth_collection
+            )
+            self.score = self._calculate_score(verdict_collection)
+            self.reason: str | None = await self._generate_reason(
```

Comment:
I'd have thought we didn't need to specify type here since it's part of the method signature?